thinc/layers/embed.py

```diff
@@ -4,12 +4,12 @@
 from .array_getitem import ints_getitem
 from ..model import Model
 from ..config import registry
-from ..types import Ints1d, Ints2d, Floats1d, Floats2d
+from ..types import Ints1d, Floats1d, Floats2d
 from ..initializers import uniform_init
 from ..util import get_width, partial
 
 
-InT = Union[Ints1d, Ints2d]
+InT = Ints1d
```

Comment:
Here the declared and documented type included Ints2d, but the backprop method appears not to support it.

Reply:
Hmm, I wonder what the original idea was behind supporting `Ints2d`. At least `scatter_add` is supposed to work on `IntsXd` so maybe this would just work for `Ints2d` as well. 

@danieldk , can you have a look as well? There's a similar issue with `hashembed`

Reply:
I think `Ints2d` is valid when `column` is given as an argument. This is used in spaCy:

https://github.com/explosion/spaCy/blob/72abf9e1021df4300d0f776b80980442247d6a9a/spacy/ml/models/tok2vec.py#L154

Reply:
I can see how the forward method of `embed` would work for `Ints1d` and `Ints2d` and how the forward method of `hashembed` would work for `IntsXd`. And in the latter case we seem to have found an example where this is the case. However, it is definitely the case that both backprop methods return `Ints1d`s that are not just declared as such, but allocated. Perhaps I'm missing some crucial piece of information about how backpropping works: I thought the output of a backprop method would always have to match the input of the corresponding forward method?

Reply:
Thanks @danieldk, I've finally got it now. I hadn't seen that the models returned by the `Embed()` and `HashEmbed()` methods are not the same as the models that the respective `forward()` and `backprop()` methods deal with. So if I've understood it correctly:

- The original version is wrong because the model returned by the layer is typed the same way as the models that `forward()` and `backprop()` deal with, while these are actually distinct models with distinct types. Although the layer as a whole can accept an `Ints2d`, the `forward()` and `backprop()` methods can't.
- The version in this PR is also wrong because it prevents the layers from accepting `Ints2d` inputs.
- The correct course of action is to alter the definition of `InT` back to what it was before, but to use `Ints1d` rather than `InT`in the `forward()` and `backprop()` methods.

Is that right?

Reply:
Yeah, sounds completely right to me!